### PR TITLE
Threads: replace enum with constexpr int and enum class

### PR DESCRIPTION
**Pull request by Rombur**
This PR replaces the `enum` in `Kokkos_Threads_Instance` with `constexpr int` and `enum class` and move the `enum class` to its own file. The PR also moves `Kokkos_Spinwait.hpp` from `Impl` to `Threads`. `spinwait_while_equal` now takes a `ThreadState` as an argument. The template parameter has been removed. The following functions, which where not used, have been removed:
 - `root_spinwait_while_equal`
 - `root_spinwait_until_equal`
 - `spinwait_until_equal`
 - `yield_until_equal`

### CODE REVIEW

File: core/src/Threads/Kokkos_Threads_Instance.cpp [hunk 1]
```diff
@@ -278,12 +278,11 @@ void ThreadsInternal::execute_sleep(ThreadsInternal &exec, const void *) {
   const int rank_rev = exec.m_pool_size - (exec.m_pool_rank + 1);
 
   for (int i = 0; i < n; ++i) {
-    Impl::spinwait_while_equal<int>(
-        exec.m_pool_base[rank_rev + (1 << i)]->m_pool_state,
-        ThreadsInternal::Active);
+    Impl::spinwait_while_equal<State>(
+        exec.m_pool_base[rank_rev + (1 << i)]->m_pool_state, State::Active);
```
Comment (dalg24) [inline core/src/Threads/Kokkos_Threads_Instance.cpp:282]:
```suggestion
    Impl::spinwait_while_equal(m_pool_base[rev_rank + (1 << i)]->m_pool_state,
                               State::Active);
```
Same comment everywhere.  I don't think there is any point on specifying the template parameters.

File: core/src/impl/Kokkos_Spinwait.hpp [hunk 2]
```diff
@@ -38,8 +38,8 @@ enum class WaitMode : int {
 void host_thread_yield(const uint32_t i, const WaitMode mode);
 
 template <typename T>
-std::enable_if_t<std::is_integral<T>::value, void> root_spinwait_while_equal(
-    T const volatile& flag, const T value) {
+std::enable_if_t<std::is_integral_v<T> || std::is_enum_v<T>, void>
+root_spinwait_while_equal(T const volatile& flag, const T value) {
```
Comment (dalg24) [inline core/src/impl/Kokkos_Spinwait.hpp:42]: Why are these templated at all?  Was it some way of avoiding circular dependencies?

Comment (Rombur) [reply core/src/impl/Kokkos_Spinwait.hpp:42]: Yeah but now that I think about it more, I could also move the `enum` outside of the class into a separate file, move `Kokkos_Spinwait` to the `Threads` backend folder, and remove the template parameter. Not sure if one solution is better than the other.

**Comment by Rombur** [comment]
I moved `Kokkos_Spinwait` to `Threads` and removed the unused spinwait/yield functions .

File: core/src/Threads/Kokkos_Threads_Spinwait.hpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_THREADS_SPINWAIT_HPP
+#define KOKKOS_THREADS_SPINWAIT_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Atomic.hpp>
```
Comment (masterleinad) [inline core/src/Threads/Kokkos_Threads_Spinwait.hpp:21]: Why do you need `atomics` here?

File: core/src/Threads/Kokkos_Threads_Spinwait.hpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+#include <Threads/Kokkos_Threads_State.hpp>
+
+#include <cstdint>
+
+#include <type_traits>
```
Comment (masterleinad) [inline core/src/Threads/Kokkos_Threads_Spinwait.hpp:27]: What do you need this header for?

File: core/src/Threads/Kokkos_Threads_Spinwait.hpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_THREADS_SPINWAIT_HPP
+#define KOKKOS_THREADS_SPINWAIT_HPP
+
+#include <Kokkos_Macros.hpp>
```
Comment (masterleinad) [inline core/src/Threads/Kokkos_Threads_Spinwait.hpp:20]: Are you using ay macros explicitly?

**Comment by Rombur** [comment]
Retest this please

**Review by dalg24 — COMMENTED** [review]
Please update the description and document what functions are being dropped because they were not used.

**Review by dalg24 — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

**Comment by dalg24** [comment]
Ignoring all the timeouts

**Pull request merged**